src/transformers/models/marian/modeling_marian.py

```diff
@@ -1085,26 +1092,78 @@ def __init__(self, config: MarianConfig):
         padding_idx, vocab_size = config.pad_token_id, config.vocab_size
         self.shared = nn.Embedding(vocab_size, config.d_model, padding_idx)
 
-        self.encoder = MarianEncoder(config, self.shared)
-        self.decoder = MarianDecoder(config, self.shared)
+        if self.config.share_encoder_decoder_embeddings:
+            encoder_embed_tokens = decoder_embed_tokens = self.shared
+        else:
+            encoder_embed_tokens = copy.deepcopy(self.shared)
```

Comment:
Can we maybe add a nice comment here why we do the cloning in this branch and why we don't do it above? E.g. if it's' not shared we want the embed_tokens to be correctly initialized, but **not** tied which is why we do the deepcopy here. Think it's worth adding a comment here